crates/rover-std/src/fs.rs

```diff
@@ -54,23 +55,70 @@ impl Fs {
         C: AsRef<[u8]>,
     {
         let path = path.as_ref();
-        let contents = str::from_utf8(contents.as_ref()).with_context(|| {
-            format!(
-                "tried to write contents to {} that was invalid UTF-8",
-                &path
-            )
-        })?;
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
-        }
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
+        tracing::info!("checking existence of parent path in '{}'", path);
+        // Grab the parent path then attempt to canonicalize it, we can't just canonicalize the
+        // entire path because that would entail the file existing, which of course it doesn't yet.
+        let mut canonical_final_path = if let Some(parent_path) = path.parent() {
+            tracing::debug!("attempting to canonicalize parent path '{}'", parent_path);
+            Self::upsert_path_exists(parent_path)
+        } else {
+            // If the has no parent, the only conceivable case is that we are looking at the root,
+            // and as that cannot be a file we should bail out at this point
+            return Err(anyhow::anyhow!("cannot write file to root or prefix {}", path).into());
+        }?;
+
+        return if let Some(file_name) = path.file_name() {
+            // Create the final version of the path we want to create
+            canonical_final_path.push(file_name);
+            tracing::debug!("final canonical path is {}", canonical_final_path);
+            // Setup a file pointer
+            let mut file = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .truncate(true)
+                .open(path)
+                .with_context(|| {
+                    format!(
+                        "tried to open {} but was unable to do so",
+                        &canonical_final_path
+                    )
+                })?;
+            tracing::info!("writing {} to disk", &canonical_final_path);
+            // Actually write the file out to where it needs to be
+            file.write(contents.as_ref())
+                .with_context(|| format!("could not write {}", &canonical_final_path))?;
+            Ok(())
+        } else {
+            // If we can't pull out a final element, then we're in a weird case but should bail
+            // out as there's nothing sensible we can do.
+            Err(anyhow::anyhow!("cannot write to a path without a final element {}", path).into())
+        };
+    }
+
```

Comment:
I think there might be a way to write this in a more top-down readable-y way (at least, I got stuck on how the `return` was placed with the nested `OK(())` and `Err(..)`, but this might just by my idiosyncrasies)

here's a quick switcheroo for how to make it read more linerally:
```
    {
        let path = path.as_ref();
        tracing::info!("checking existence of parent path in '{}'", path);
        // Grab the parent path then attempt to canonicalize it, we can't just canonicalize the
        // entire path because that would entail the file existing, which of course it doesn't yet.
        //
        // If we can't pull out a final element, then we're in a weird case but should bail
        // out as there's nothing sensible we can do.
        let file_name = path.file_name().ok_or(anyhow::anyhow!(
            "cannot write to a path without a final element {path}"
        ))?;

        let mut canonical_final_path = path
            .parent()
            .map(Self::upsert_path_exists)
            // If the has no parent, the only conceivable case is that we are looking at the root,
            // and as that cannot be a file we should bail out at this point
            .ok_or(anyhow::anyhow!(
                "cannot write file to root or prefix {path}"
            ))??;

        canonical_final_path.push(file_name);

        tracing::debug!("final canonical path is {}", canonical_final_path);

        let mut file = OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)
            .with_context(|| {
                format!(
                    "tried to open {} but was unable to do so",
                    &canonical_final_path
                )
            })?;

        tracing::info!("writing {} to disk", &canonical_final_path);

        // Actually write the file out to where it needs to be
        file.write(contents.as_ref())
            .with_context(|| format!("could not write {}", &canonical_final_path))?;

        Ok(())
    }
```

I moved the `debug` log for the attempt to canonicalize into the `upsert_path_exists` fn

Reply:
This is a great shout, thanks!